src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,575 @@
+    }
+
+    pos = working_buffer.end();
+    auto ret_off = file_reader->seek(new_pos);
+    return ret_off;
+}
+
+off_t RemoteReadBuffer::getPosition()
+{
+    return file_reader->getOffset() - available();
+}
+
+RemoteReadBufferCache::RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache::~RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache & RemoteReadBufferCache::instance()
+{
+    static RemoteReadBufferCache instance;
+    return instance;
+}
+
+void RemoteReadBufferCache::recoverCachedFilesMetaData(
+    const fs::path & current_path,
+    size_t current_depth,
+    size_t max_depth)
+{
+    if (current_depth >= max_depth)
+    {
+        for (auto const & dir : fs::directory_iterator{current_path})
+        {
+            String path = dir.path();
+            auto cache_controller = RemoteCacheController::recover(path);
+            if (!cache_controller)
+                continue;
+            auto & cell = caches[path];
+            cell.cache_controller = cache_controller;
+            cell.key_iterator = keys.insert(keys.end(), path);
+        }
+        return;
+    }
+
+    for (auto const & dir : fs::directory_iterator{current_path})
+    {
+        recoverCachedFilesMetaData(dir.path(), current_depth + 1, max_depth);
+    }
+}
+
+void RemoteReadBufferCache::recoverTask()
+{
+    std::lock_guard lock(mutex);
+    recoverCachedFilesMetaData(root_dir, 1, 2);
+    initialized = true;
+    LOG_TRACE(log, "Recovered from directory:{}", root_dir);
+}
+
+void RemoteReadBufferCache::initOnce(
+    ContextPtr context,
+    const String & root_dir_, size_t limit_size_, size_t bytes_read_before_flush_)
+{
+    LOG_INFO(
+        log, "Initializing local cache for remote data sources. Local cache root path: {}, cache size limit: {}", root_dir_, limit_size_);
+    root_dir = root_dir_;
+    limit_size = limit_size_;
+    local_cache_bytes_read_before_flush = bytes_read_before_flush_;
+
+    /// create if root_dir not exists
+    if (!fs::exists(fs::path(root_dir) / ""))
+    {
+        std::error_code ec;
+        bool success = fs::create_directories(fs::path(root_dir) / "", ec);
+        if (!success)
+            throw Exception(
+                ErrorCodes::CANNOT_CREATE_DIRECTORY, "Failed to create directories, error code:{} reason:{}", ec.value(), ec.message());
+    }
+
+    recover_task_holder = context->getSchedulePool().createTask("recover local cache meta data for remote files", [this]{ recoverTask(); });
+    recover_task_holder->activateAndSchedule();
+}
+
+String RemoteReadBufferCache::calculateLocalPath(const RemoteFileMetadata & meta) const
+{
+    String full_path = meta.schema + ":" + meta.cluster + ":" + meta.remote_path;
+    UInt128 hashcode = sipHash128(full_path.c_str(), full_path.size());
+    String hashcode_str = getHexUIntLowercase(hashcode);
+    return fs::path(root_dir) / hashcode_str.substr(0, 3) / hashcode_str;
+}
+
+std::pair<std::shared_ptr<LocalCachedFileReader>, RemoteReadBufferCacheError>
```

Comment:
Let's substitute `std::shared_ptr<LocalCachedFileReader>` with `LocalCacheFileReaderPtr = std::shared_ptr<LocalCachedFileReader>`

Also is it necessary to process error when returning from `createReader`? May be just throw proper exception if needed here, and if we want to silently quit, then return `nullptr` for reader and this will be an indicator for that?

Reply:
done